public/js/twiliochat.js

```diff
@@ -150,40 +154,60 @@ var twiliochat = (function() {
     }
   };
 
+  function initChannel(channel) {
+    console.log('Initialized channel ' + channel.friendlyName);
+    return tc.messagingClient.getChannelBySid(channel.sid);
+  }
+
+  function joinChannel(_channel) {
+    return _channel.join()
+      .then(function(joinedChannel) {
+        console.log('Joined channel ' + joinedChannel.friendlyName);
+        updateChannelUI(_channel);
+        tc.currentChannel = _channel;
+        tc.loadMessages();
+        return joinedChannel;
+      });
+  }
+
+  function initChannelEvents() {
+    console.log(tc.currentChannel.friendlyName + ' ready.');
+    tc.currentChannel.on('messageAdded', tc.addMessageToList);
+    tc.currentChannel.on('typingStarted', showTypingStarted);
+    tc.currentChannel.on('typingEnded', hideTypingStarted);
+    tc.currentChannel.on('memberJoined', notifyMemberJoined);
+    tc.currentChannel.on('memberLeft', notifyMemberLeft);
+    $inputText.prop('disabled', false).focus();
+  }
+
   function setupChannel(channel) {
-    // Join the channel
-    channel.join().then(function(joinedChannel) {
-      console.log('Joined channel ' + joinedChannel.friendlyName);
-      leaveCurrentChannel();
-      updateChannelUI(channel);
-      tc.currentChannel = channel;
-      tc.loadMessages();
-      channel.on('messageAdded', tc.addMessageToList);
-      channel.on('typingStarted', showTypingStarted);
-      channel.on('typingEnded', hideTypingStarted);
-      channel.on('memberJoined', notifyMemberJoined);
-      channel.on('memberLeft', notifyMemberLeft);
-      $inputText.prop('disabled', false).focus();
-      tc.$messageList.text('');
-    });
+    return leaveCurrentChannel()
+      .then(initChannel(channel)
```

Comment:
I would avoid nesting chains of promises when possible:
```
function setupChannel(channel) {
    return leaveCurrentChannel()
      .then(initChannel(channel))
      .then(function(_channel) {
          return joinChannel(_channel); // you need the return here, otherwise the next promise will run too soon
      })
      .then(initChannelEvents);
  }
```